src/Modules/OrchardCore.Commerce.Payment/Controllers/PaymentBaseController.cs

```diff
@@ -0,0 +1,159 @@
+using Microsoft.AspNetCore.Http.Extensions;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using OrchardCore.Commerce.Payment.Constants;
+using OrchardCore.Commerce.Payment.ViewModels;
+using OrchardCore.DisplayManagement.Notify;
+using OrchardCore.Mvc.Core.Utilities;
+using System;
+using System.Threading.Tasks;
+
+namespace OrchardCore.Commerce.Payment.Controllers;
+public abstract class PaymentBaseController : Controller
+{
+    private readonly INotifier _notifier;
+    private readonly ILogger _logger;
+    protected PaymentBaseController(INotifier notifier, ILogger logger)
+    {
+        _notifier = notifier;
+        _logger = logger;
+    }
+
+    public async Task<IActionResult> ProduceActionResultAsync(PaidStatusViewModel paidStatusViewModel)
+    {
+        if (paidStatusViewModel.ShowMessage != null)
+        {
+            switch (paidStatusViewModel.Status)
+            {
+                case PaidStatus.Succeeded:
+                    await _notifier.SuccessAsync(paidStatusViewModel.ShowMessage);
+                    break;
+                case PaidStatus.Failed:
+                    await LogAndNotifyFailedAsync(paidStatusViewModel);
+                    break;
+                case PaidStatus.NotFound:
+                    await LogAndNotifyWarningAsync(paidStatusViewModel);
+                    break;
+                case PaidStatus.NotThingToDo:
+                case PaidStatus.WaitingStripe:
+                case PaidStatus.WaitingPayment:
+                    await LogAndNotifyInformationAsync(paidStatusViewModel);
+                    break;
+                default:
+                    await LogAndNotifyFailedAsync(paidStatusViewModel);
+                    break;
+            }
+        }
+
+        return paidStatusViewModel.Status switch
+        {
+            PaidStatus.Succeeded => RedirectToActionWithParams<PaymentController>(
+                nameof(PaymentController.Success),
+                FeatureIds.Area,
+                string.Empty,
+                orderId: paidStatusViewModel.Content?.ContentItemId),
+
+            PaidStatus.Failed => RedirectToActionWithParams<PaymentController>(
+                nameof(PaymentController.Index),
+                FeatureIds.Payment),
+
+            PaidStatus.NotFound => NotFound(),
+
+            PaidStatus.NotThingToDo => this.RedirectToContentDisplay(paidStatusViewModel.Content),
+
+            PaidStatus.WaitingStripe => RedirectToActionWithNames(
+                "PaymentConfirmationMiddleware",
+                "OrchardCore.Commerce.Payment.Stripe",
+                "Stripe"),
```

Comment:
we cannot assign .Url in advance, because for minimal api , the Url will be different from mvc.
for the string issue, i had tried that, but there was an issue for project reference. I saw in many places of orchardcore , if there are similar issues, use string to avoid them.

Reply:
You can have two separate properties for MVC URL and API URL in the view-model. The goal is that the `StripePaymentProvider` should set its own URLs instead of the `PaymentBaseController`. This way project reference won't be a problem either.